File: src/ZoneCodeGenerator/Interface/PrettyPrinter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using ZoneCodeGenerator.Domain;
using ZoneCodeGenerator.Persistence;

namespace ZoneCodeGenerator.Interface
{
    class PrettyPrinter
    {
        private readonly IDataRepository dataRepository;

        public PrettyPrinter(IDataRepository dataRepository)
        {
            this.dataRepository = dataRepository;
        }

        public void PrintEnums()
        {
            var allEnums = dataRepository.GetAllEnums();
            var dataTypeEnums = allEnums.ToList();

            Console.WriteLine($"{dataTypeEnums.Count} enums:");

            foreach(var _enum in dataTypeEnums)
            {
                Console.WriteLine($" Name: {_enum.FullName}");
                Console.WriteLine($" Alignment: {_enum.Alignment}");
                Console.WriteLine($" Size: {_enum.Size}");

                foreach(var member in _enum.Members)
                {
                    Console.WriteLine($"  {member.Name} = {member.Value}");
                }

                Console.WriteLine();
            }
        }

        private static void PrintVariablePointerToArray(Variable variable)
        {
            var arraySize = new List<int>();
            var pointerDepth = 0;
            var referenceOffset = 0;

            for (; referenceOffset < variable.VariableType.References.Count; referenceOffset++)
            {
                if (!(variable.VariableType.References[referenceOffset] is ReferenceTypePointer))
                    break;
                
                pointerDepth++;
            }

            for (; referenceOffset < variable.VariableType.References.Count; referenceOffset++)
            {
                if (!(variable.VariableType.References[referenceOffset] is ReferenceTypeArray array))
                    throw new Exception("Expected the rest of the references to be array references");
                
                arraySize.Add(array.ArraySize);
            }

            var memberBuilder = new StringBuilder();
            memberBuilder.Append("  ");
            memberBuilder.Append(variable.VariableType.Type.FullName);
            memberBuilder.Append('(');
            memberBuilder.Append('*', pointerDepth);
            memberBuilder.Append(variable.Name);
            memberBuilder.Append(')');

            foreach (var array in arraySize)
            {
                memberBuilder.Append('[');
                memberBuilder.Append(array);
                memberBuilder.Append(']');
            }

            Console.WriteLine(memberBuilder.ToString());
        }

        private static void PrintVariableArrayOfPointers(Variable variable)
        {
            var arraySize = new List<int>();
            var pointerDepth = 0;
            var referenceOffset = 0;

            for (; referenceOffset < variable.VariableType.References.Count; referenceOffset++)
            {
                if (!(variable.VariableType.References[referenceOffset] is ReferenceTypeArray array))
                    break;

                arraySize.Add(array.ArraySize);
            }

            for (; referenceOffset < variable.VariableType.References.Count; referenceOffset++)
            {
                if (!(variable.VariableType.References[referenceOffset] is ReferenceTypePointer))
                    throw new Exception("Expected the rest of the references to be pointer references");

                pointerDepth++;
            }

            var memberBuilder = new StringBuilder();
            memberBuilder.Append("  ");
            memberBuilder.Append(variable.VariableType.Type.FullName);
            memberBuilder.Append('*', pointerDepth);
            memberBuilder.Append(" ");
            memberBuilder.Append(variable.Name);

            foreach (var array in arraySize)
            {
                memberBuilder.Append('[');
                memberBuilder.Append(array);
                memberBuilder.Append(']');
            }

            Console.WriteLine(memberBuilder.ToString());
        }

        private static void PrintVariable(Variable variable)
        {
            if (variable.VariableType.References.Count == 0)
            {
                Console.WriteLine($"  {variable.VariableType.Type.FullName} {variable.Name}");
            }
            else
            {
                if (variable.VariableType.References[0] is ReferenceTypePointer &&
                    variable.VariableType.References.OfType<ReferenceTypeArray>().Any())
                {
                    PrintVariablePointerToArray(variable);
                }
                else
                {
                    PrintVariableArrayOfPointers(variable);
                }
            }
        }

        public void PrintStructs()
        {
            var allStructs = dataRepository.GetAllStructs();
            var dataTypeStructs = allStructs.ToList();

            Console.WriteLine($"{dataTypeStructs.Count} structs:");

            foreach(var _struct in dataTypeStructs)
            {
                Console.WriteLine($" Name: {_struct.FullName}");
                Console.WriteLine($" Alignment: {_struct.Alignment}");
                Console.WriteLine($" Size: {_struct.Size}");

                foreach(var member in _struct.Members)
                    PrintVariable(member);

                Console.WriteLine();
            }

        }

        public void PrintUnions()
        {
            var allUnions = dataRepository.GetAllUnions();
            var dataTypeUnions = allUnions.ToList();

            Console.WriteLine($"{dataTypeUnions.Count} unions:");

            foreach(var union in dataTypeUnions)
            {
                Console.WriteLine($" Name: {union.FullName}");
                Console.WriteLine($" Alignment: {union.Alignment}");
                Console.WriteLine($" Size: {union.Size}");

                foreach(var member in union.Members)
                    PrintVariable(member);

                Console.WriteLine();
            }
        }

        private static void PrintTypedefPointerToArray(DataTypeTypedef typedef)
        {
            var arraySize = new List<int>();
            var pointerDepth = 0;
            var referenceOffset = 0;

            for (; referenceOffset < typedef.TypeDefinition.References.Count; referenceOffset++)
            {
                if (!(typedef.TypeDefinition.References[referenceOffset] is ReferenceTypePointer))
                    break;
                
                pointerDepth++;
            }

            for (; referenceOffset < typedef.TypeDefinition.References.Count; referenceOffset++)
            {
                if (!(typedef.TypeDefinition.References[referenceOffset] is ReferenceTypeArray array))
                    throw new Exception("Expected the rest of the references to be array references");
                
                arraySize.Add(array.ArraySize);
            }

            var memberBuilder = new StringBuilder();
            memberBuilder.Append("  ");
            memberBuilder.Append(typedef.TypeDefinition.Type.FullName);
            memberBuilder.Append('(');
            memberBuilder.Append('*', pointerDepth);
            memberBuilder.Append(')');

            foreach (var array in arraySize)
            {
                memberBuilder.Append('[');
                memberBuilder.Append(array);
                memberBuilder.Append(']');
            }

            Console.WriteLine(memberBuilder.ToString());
        }

        private static void PrintTypedefArrayOfPointers(DataTypeTypedef typedef)
        {
            var arraySize = new List<int>();
            var pointerDepth = 0;
            var referenceOffset = 0;

            for (; referenceOffset < typedef.TypeDefinition.References.Count; referenceOffset++)
            {
                if (!(typedef.TypeDefinition.References[referenceOffset] is ReferenceTypeArray array))
                    break;

                arraySize.Add(array.ArraySize);
            }

            for (; referenceOffset < typedef.TypeDefinition.References.Count; referenceOffset++)
            {
                if (!(typedef.TypeDefinition.References[referenceOffset] is ReferenceTypePointer))
                    throw new Exception("Expected the rest of the references to be pointer references");

                pointerDepth++;
            }

            var memberBuilder = new StringBuilder();
            memberBuilder.Append("  ");
            memberBuilder.Append(typedef.TypeDefinition.Type.FullName);
            memberBuilder.Append('*', pointerDepth);

            foreach (var array in arraySize)
            {
                memberBuilder.Append('[');
                memberBuilder.Append(array);
                memberBuilder.Append(']');
            }

            Console.WriteLine(memberBuilder.ToString());
        }

        public void PrintTypedefs()
        {
            var allTypedefs = dataRepository.GetAllTypedefs();
            var dataTypeTypedefs = allTypedefs.ToList();

            Console.WriteLine($"{dataTypeTypedefs.Count} typedefs:");

            foreach (var typeDef in dataTypeTypedefs)
            {
                Console.WriteLine($" Name: {typeDef.FullName}");
                Console.WriteLine($" Alignment: {typeDef.Alignment}");
                Console.WriteLine($" Size: {typeDef.Size}");

                if (typeDef.TypeDefinition.References.Count == 0)
                {
                    Console.WriteLine($"  {typeDef.TypeDefinition.Type.FullName}");
                }
                else
                {
                    if (typeDef.TypeDefinition.References[0] is ReferenceTypePointer &&
                        typeDef.TypeDefinition.References.OfType<ReferenceTypeArray>().Any())
                    {
                        PrintTypedefPointerToArray(typeDef);
                    }
                    else
                    {
                        PrintTypedefArrayOfPointers(typeDef);
                    }
                }

                Console.WriteLine();
            }

        }

        public void PrintAll()
        {
            PrintEnums();

            Console.WriteLine();
            Console.WriteLine("==========================================================================================================");
            Console.WriteLine();

            PrintStructs();

            Console.WriteLine();
            Console.WriteLine("==========================================================================================================");
            Console.WriteLine();

            PrintUnions();

            Console.WriteLine();
            Console.WriteLine("==========================================================================================================");
            Console.WriteLine();

            PrintTypedefs();
        }
    }
}

File: src/ZoneCodeGeneratorLib/Printing/PrettyPrinter.cpp
#include "PrettyPrinter.h"

#include "Domain/Definition/ArrayDeclarationModifier.h"
#include "Utils/NamespaceBuilder.h"

#include <algorithm>
#include <format>
#include <iostream>

PrettyPrinter::PrettyPrinter(std::ostream& stream, const IDataRepository* repository)
    : m_stream(stream),
      m_repository(repository)
{
}

void PrettyPrinter::PrintSeparator() const
{
    m_stream << "\n==========================================================================================================\n\n";
}

void PrettyPrinter::PrintVariablePointerToArray(Variable* variable) const
{
    const auto& declarationModifiers = variable->m_type_declaration->m_declaration_modifiers;

    std::vector<int> arraySize;
    auto pointerDepth = 0;
    auto modifierOffset = 0u;

    for (; modifierOffset < declarationModifiers.size(); modifierOffset++)
    {
        const auto& modifier = declarationModifiers[modifierOffset];
        if (modifier->GetType() != DeclarationModifierType::POINTER)
            break;

        pointerDepth++;
    }

    for (; modifierOffset < declarationModifiers.size(); modifierOffset++)
    {
        const auto& modifier = declarationModifiers[modifierOffset];
        if (modifier->GetType() != DeclarationModifierType::ARRAY)
            break;

        arraySize.push_back(dynamic_cast<ArrayDeclarationModifier*>(modifier.get())->m_size);
    }

    m_stream << std::format("  {}(", variable->m_type_declaration->m_type->GetFullName());
    for (auto p = 0; p < pointerDepth; p++)
        m_stream << '*';
    m_stream << std::format("{})", variable->m_name);
    for (auto size : arraySize)
        m_stream << std::format("[{}]", size);
    m_stream << "\n";
}

void PrettyPrinter::PrintVariableArrayOfPointers(Variable* variable) const
{
    const auto& declarationModifiers = variable->m_type_declaration->m_declaration_modifiers;

    std::vector<int> arraySize;
    auto pointerDepth = 0;
    auto modifierOffset = 0u;

    for (; modifierOffset < declarationModifiers.size(); modifierOffset++)
    {
        const auto& modifier = declarationModifiers[modifierOffset];
        if (modifier->GetType() != DeclarationModifierType::ARRAY)
            break;

        arraySize.push_back(dynamic_cast<ArrayDeclarationModifier*>(modifier.get())->m_size);
    }

    for (; modifierOffset < declarationModifiers.size(); modifierOffset++)
    {
        const auto& modifier = declarationModifiers[modifierOffset];
        if (modifier->GetType() != DeclarationModifierType::POINTER)
            break;

        pointerDepth++;
    }

    m_stream << std::format("  {}", variable->m_type_declaration->m_type->GetFullName());
    for (auto p = 0; p < pointerDepth; p++)
        m_stream << '*';
    m_stream << std::format(" {}", variable->m_name);
    for (auto size : arraySize)
        m_stream << std::format("[{}]", size);
    m_stream << "\n";
}

void PrettyPrinter::PrintVariable(Variable* variable) const
{
    const auto& declarationModifiers = variable->m_type_declaration->m_declaration_modifiers;
    if (declarationModifiers.empty())
    {
        m_stream << std::format("  {} {}\n", variable->m_type_declaration->m_type->GetFullName(), variable->m_name);
    }
    else
    {
        if (declarationModifiers[0]->GetType() == DeclarationModifierType::POINTER
            && std::ranges::any_of(declarationModifiers,
                                   [](const std::unique_ptr<DeclarationModifier>& modifier)
                                   {
                                       return modifier->GetType() == DeclarationModifierType::ARRAY;
                                   }))
        {
            PrintVariablePointerToArray(variable);
        }
        else
        {
            PrintVariableArrayOfPointers(variable);
        }
    }
}

void PrettyPrinter::PrintTypedefPointerToArray(TypedefDefinition* typedefDefinition) const
{
    const auto& declarationModifiers = typedefDefinition->m_type_declaration->m_declaration_modifiers;

    std::vector<int> arraySize;
    auto pointerDepth = 0;
    auto modifierOffset = 0u;

    for (; modifierOffset < declarationModifiers.size(); modifierOffset++)
    {
        const auto& modifier = declarationModifiers[modifierOffset];
        if (modifier->GetType() != DeclarationModifierType::POINTER)
            break;

        pointerDepth++;
    }

    for (; modifierOffset < declarationModifiers.size(); modifierOffset++)
    {
        const auto& modifier = declarationModifiers[modifierOffset];
        if (modifier->GetType() != DeclarationModifierType::ARRAY)
            break;

        arraySize.push_back(dynamic_cast<ArrayDeclarationModifier*>(modifier.get())->m_size);
    }

    m_stream << std::format("  {}(", typedefDefinition->m_type_declaration->m_type->GetFullName());
    for (auto p = 0; p < pointerDepth; p++)
        m_stream << '*';
    m_stream << std::format("{})", typedefDefinition->m_name);
    for (auto size : arraySize)
        m_stream << std::format("[{}]", size);
    m_stream << "\n";
}

void PrettyPrinter::PrintTypedefArrayOfPointers(TypedefDefinition* typedefDefinition) const
{
    const auto& declarationModifiers = typedefDefinition->m_type_declaration->m_declaration_modifiers;

    std::vector<int> arraySize;
    auto pointerDepth = 0;
    auto modifierOffset = 0u;

    for (; modifierOffset < declarationModifiers.size(); modifierOffset++)
    {
        const auto& modifier = declarationModifiers[modifierOffset];
        if (modifier->GetType() != DeclarationModifierType::ARRAY)
            break;

        arraySize.push_back(dynamic_cast<ArrayDeclarationModifier*>(modifier.get())->m_size);
    }

    for (; modifierOffset < declarationModifiers.size(); modifierOffset++)
    {
        const auto& modifier = declarationModifiers[modifierOffset];
        if (modifier->GetType() != DeclarationModifierType::POINTER)
            break;

        pointerDepth++;
    }

    m_stream << std::format("  {}", typedefDefinition->m_type_declaration->m_type->GetFullName());
    for (auto p = 0; p < pointerDepth; p++)
        m_stream << '*';
    m_stream << std::format(" {}", typedefDefinition->m_name);
    for (auto size : arraySize)
        m_stream << std::format("[{}]", size);
    m_stream << "\n";
}

void PrettyPrinter::PrintEnums() const
{
    const auto& allEnums = m_repository->GetAllEnums();
    m_stream << std::format("{} enums:\n", allEnums.size());

    for (auto* enumDefinition : allEnums)
    {
        m_stream << std::format(" Name: {}\n", enumDefinition->GetFullName());
        m_stream << std::format(" Alignment: {}\n", enumDefinition->GetAlignment());
        m_stream << std::format(" Size: {}\n", enumDefinition->GetSize());

        for (const auto& enumMember : enumDefinition->m_members)
        {
            m_stream << std::format("  {} = {}\n", enumMember->m_name, enumMember->m_value);
        }

        m_stream << "\n";
    }
}

void PrettyPrinter::PrintStructs() const
{
    const auto& allStructs = m_repository->GetAllStructs();
    m_stream << std::format("{} structs:\n", allStructs.size());

    for (auto* structDefinition : allStructs)
    {
        m_stream << std::format(" Name: {}\n", structDefinition->GetFullName());
        m_stream << std::format(" Alignment: {}\n", structDefinition->GetAlignment());
        m_stream << std::format(" Size: {}\n", structDefinition->GetSize());

        for (const auto& variable : structDefinition->m_members)
        {
            PrintVariable(variable.get());
        }

        m_stream << "\n";
    }
}

void PrettyPrinter::PrintUnions() const
{
    const auto& allUnions = m_repository->GetAllUnions();
    m_stream << std::format("{} unions:\n", allUnions.size());

    for (auto* unionDefinition : allUnions)
    {
        m_stream << std::format(" Name: {}\n", unionDefinition->GetFullName());
        m_stream << std::format(" Alignment: {}\n", unionDefinition->GetAlignment());
        m_stream << std::format(" Size: {}\n", unionDefinition->GetSize());

        for (const auto& variable : unionDefinition->m_members)
        {
            PrintVariable(variable.get());
        }

        m_stream << "\n";
    }
}

void PrettyPrinter::PrintTypedefs() const
{
    const auto& allTypedefs = m_repository->GetAllTypedefs();
    m_stream << std::format("{} typedefs:\n", allTypedefs.size());

    for (auto* typedefDefinition : allTypedefs)
    {
        m_stream << std::format(" Name: {}\n", typedefDefinition->GetFullName());
        m_stream << std::format(" Alignment: {}\n", typedefDefinition->GetAlignment());
        m_stream << std::format(" Size: {}\n", typedefDefinition->GetSize());

        const auto& declarationModifiers = typedefDefinition->m_type_declaration->m_declaration_modifiers;
        if (declarationModifiers.empty())
        {
            m_stream << std::format("  {}\n", typedefDefinition->m_type_declaration->m_type->GetFullName());
        }
        else
        {
            if (declarationModifiers[0]->GetType() == DeclarationModifierType::POINTER
                && std::ranges::any_of(declarationModifiers,
                                       [](const std::unique_ptr<DeclarationModifier>& modifier)
                                       {
                                           return modifier->GetType() == DeclarationModifierType::ARRAY;
                                       }))
            {
                PrintTypedefPointerToArray(typedefDefinition);
            }
            else
            {
                PrintTypedefArrayOfPointers(typedefDefinition);
            }
        }

        m_stream << "\n";
    }
}

void PrettyPrinter::PrintAll() const
{
    PrintEnums();

    PrintSeparator();

    PrintStructs();

    PrintSeparator();

    PrintUnions();

    PrintSeparator();

    PrintTypedefs();
}
